core/src/solve.jl

```diff
@@ -478,34 +479,76 @@ function continuous_control!(
 
         listened_node_id = listen_node_id[i]
         _, listened_node_idx = id_index(basin.node_id, listened_node_id)
+        storage_listened_basin = u.storage[listened_node_idx]
+        phi = storage_listened_basin < 10.0 ? storage_listened_basin / 10.0 : 1.0
 
         flow_rate = 0.0
 
         if !isnan(proportional[i])
-            flow_rate += proportional[i] * error[i]
+            flow_rate += phi * proportional[i] * error[i]
+        end
+
+        if !isnan(integral[i])
+            # coefficient * current value of integral
+            flow_rate += phi * integral[i] * integral_value[i]
         end
 
         if !isnan(derivative[i])
             # dlevel/dstorage = 1/area
             area = basin.current_area[listened_node_idx]
-
-            error_deriv = -dstorage[listened_node_idx] / area
-            flow_rate += derivative[i] * error_deriv
-        end
-
-        if !isnan(integral[i])
-            # coefficient * current value of integral
-            flow_rate += integral[i] * integral_value[i]
+            dlevel = du.storage[listened_node_idx] / area
+            dtarget_level = 0.0
+            du_listened_basin_old = du.storage[listened_node_idx]
+            du_listened_basin =
+                (du_listened_basin_old - flow_rate - phi * derivative[i] * dtarget_level) /
+                (1 - phi * derivative[i] / area)
+            flow_rate = du_listened_basin_old - du_listened_basin
         end
 
         # Clip values outside pump flow rate bounds
-        flow_rate = max(flow_rate, min_flow_rate[i])
+        was_clipped = false
+
+        if flow_rate < min_flow_rate[i]
+            was_clipped = true
+            flow_rate = min_flow_rate[i]
+        end
 
         if !isnan(max_flow_rate[i])
-            flow_rate = min(flow_rate, max_flow_rate[i])
+            if flow_rate > max_flow_rate[i]
+                was_clipped = true
+                flow_rate = max_flow_rate[i]
+            end
         end
```

Comment:
I don't see `was_clipped` being used?

It might be easier to use `clamp(x, lo, hi)` here. That also doesn't need special NaN handling:
```julia
julia> clamp(-4.0, 0.0, NaN)
0.0
```

Reply:
Nice, that indeed suffices here. In the Jac computations I do need to now whether the flow rate was actually changed here.